lib/trees/avl_tree.dart

```diff
@@ -348,88 +364,144 @@ class AvlTree<V extends Comparable> extends BinaryTreeADT<AvlNode<V>, V> {
+
       if (_isShorter) {
         node = _dUpdateLeftBalanceFactor(node);
       }
-    } else if (node.value.compareTo(value) < 0) {
+    } else if (node.value!.compareTo(value) < 0) {
       node.right = _delete(node.right, value);
+
       if (_isShorter) {
         node = _dUpdateRightBalanceFactor(node);
       }
     }
 
-    // [node] with value found.
+    // [node] with [value] found.
     else {
       // [node] has two children.
       if (node.left != null && node.right != null) {
         // Successor to the node is the next inOrder node.
         var successor = node.right;
-        while (successor.left != null) {
+        while (successor!.left != null) {
           successor = successor.left;
         }
+        // Copy [successor]'s [value] to [node] and now delete the [successor].
+        //
+        // This step shifts the deletion to the successor (having single or no
+        // child; easier to delete) from the node (having two children; much
+        // harder to delete).
+        // Similarly, instead of successor, the predecessor's value can also be
+        // copied and deletion then shifted to predecessor (having single or
+        // no child).
         node.value = successor.value;
-        node.right = _delete(node.right, successor.value);
+        node.right = _delete(node.right, successor.value!);
+
         if (_isShorter) {
+          // [successor] is always in right subtree of the [node].
           node = _dUpdateRightBalanceFactor(node);
         }
       }
 
-      // [node] only has left child.
+      // [node] only has [left] child.
       else if (node.left != null) {
+        // Simply replace [node] with it's child.
         node = node.left;
       }
 
-      // [node] only has right child.
+      // [node] only has [right] child.
       else if (node.right != null) {
+        // Simply replace [node] with it's child.
         node = node.right;
       }
 
       // Childless node.
       else {
         node = null;
       }
+
+      // Mark that this subtree is now shorter because of the deletion so, if
+      // need be, it can be balanced out while unwinding the recursion.
       _isShorter = true;
     }
     return node;
   }
 
-  /// Updates [balanceFactor] when deletion is done from left subtree of [node].
-  AvlNode _dUpdateLeftBalanceFactor(AvlNode node) {
+  /// Updates [balanceFactor] of the [node] when deletion is done from it's left
+  /// subtree.\
+  AvlNode<V> _dUpdateLeftBalanceFactor(AvlNode<V> node) {
     switch (node.balanceFactor) {
 
-      // node was balanced.
+      // [node] was balanced before deletion. (Both left and right subtree had
+      // the same height.)
+      //          N
+      //         / \
+      //        ** **
       case 0:
-        // node is right heavy now.
+        // [node] is right heavy now.
+        //          N
+        //         / \
+        //        *  **
         node.balanceFactor = -1;
         _isShorter = false;
         break;
 
```

Comment:
@code-shoily Do you think the added comment helps or it's just cluttering?

Reply:
No it does help.